src/ClassDiscovery.php

```diff
@@ -2,126 +2,117 @@
+                    }
+                }
 
-        return self::$puliFactory;
-    }
+                // save the result for later use
+                self::storeInCache($type, $candidate);
 
-    /**
-     * Sets the Puli factory.
-     *
-     * @param object $puliFactory
-     */
-    public static function setPuliFactory($puliFactory)
-    {
-        if (!is_callable([$puliFactory, 'createRepository']) || !is_callable([$puliFactory, 'createDiscovery'])) {
-            throw new \InvalidArgumentException('The Puli Factory must expose a repository and a discovery');
+                return $candidate['class'];
+            }
         }
 
-        self::$puliFactory = $puliFactory;
-        self::$puliDiscovery = null;
+        throw new DiscoveryFailedException('Could not find resource using any discovery strategy', $exceptions);
     }
 
     /**
-     * Resets the factory.
-     */
-    public static function resetPuliFactory()
-    {
-        self::$puliFactory = null;
-        self::$puliDiscovery = null;
-    }
-
-    /**
-     * Returns the Puli discovery layer.
+     * Get a value from cache.
+     *
+     * @param string $type
      *
-     * @return Discovery
+     * @return string|null
      */
-    public static function getPuliDiscovery()
+    private static function getFromCache($type)
     {
-        if (!isset(self::$puliDiscovery)) {
-            $factory = self::getPuliFactory();
-            $repository = $factory->createRepository();
+        if (!isset(self::$cache[$type])) {
+            return;
+        }
 
-            self::$puliDiscovery = $factory->createDiscovery($repository);
+        $candidate = self::$cache[$type];
+        if (!self::evaluateCondition($candidate['condition'])) {
+            return;
         }
 
-        return self::$puliDiscovery;
+        return $candidate['class'];
     }
 
     /**
-     * Finds a class.
-     *
-     * @param $type
-     *
-     * @return string
+     * Store a value in cache.
      *
-     * @throws NotFoundException
+     * @param string $type
+     * @param string $class
      */
-    public static function findOneByType($type)
+    private static function storeInCache($type, $class)
     {
-        $bindings = self::getPuliDiscovery()->findBindings($type);
-
-        foreach ($bindings as $binding) {
-            if ($binding->hasParameterValue('depends')) {
-                $dependency = $binding->getParameterValue('depends');
-
-                if (!self::evaluateCondition($dependency)) {
-                    continue;
-                }
-            }
-
-            // TODO: check class binding
-            return $binding->getClassName();
-        }
-
-        throw new NotFoundException(sprintf('Resource of type "%s" not found', $type));
+        self::$cache[$type] = $class;
     }
 
     /**
-     * Finds a resource.
-     *
-     * @return object
+     * Clear the cache.
      */
-    public static function find()
+    public static function clearCache()
```

Comment:
Good, I should move this to the Discovery classes? Though, I created this because I had to make the tests work. The only real use case I can see is where you use discovery and then add strategies yourself. Then you have to clear the cache to make the new strategies to work.